src/main.cpp

```diff
@@ -90,87 +91,197 @@ namespace
+							const adm::group group{editor.at("name").get_ref<const std::string&>()};
+
+							if (adm::server::user_is_member_of_group(*rei.rsComm, group, user)) {
+								return CODE(RULE_ENGINE_CONTINUE);
+							}
+						}
+						else if (type == "user") {
+							if (editor.at("name").get_ref<const std::string&>() ==
+							    adm::server::local_unique_name(*rei.rsComm, user)) {
+								return CODE(RULE_ENGINE_CONTINUE);
+							}
+						}
+					}
+
+					// At this point, the user is not an administrator and they aren't a member of
+					// the editors list. Therefore, we return an error because the user is attempting to
+					// modify metadata in a guarded namespace.
+					return ERROR(CAT_INSUFFICIENT_PRIVILEGE_LEVEL, "User is not allowed to modify metadata");
+				}
+			}
+		}
+		catch (const json::exception&) {
+			// clang-format off
             log_re::error({{"log_message", "Unexpected JSON access or type error."},
                            {"rule_engine_plugin", "metadata_guard"}});
             // clang-format on
-        }
-        catch (const std::exception& e) {
-            log_re::error({{"log_message", e.what()}, {"rule_engine_plugin", "metadata_guard"}});
-        }
-
-        return CODE(RULE_ENGINE_CONTINUE);
-    }
+		}
+		catch (const std::exception& e) {
+			log_re::error({{"log_message", e.what()}, {"rule_engine_plugin", "metadata_guard"}});
+		}
+		return CODE(RULE_ENGINE_CONTINUE);
+	}
+
+	auto pep_atomic_metadata_handler(std::list<boost::any>& _rule_arguments, irods::callback _effect_handler)
+		-> irods::error
+	{
+		auto is_input_valid = [](const bytesBuf_t* _input) -> std::tuple<bool, std::string> {
+			if (!_input) {
+				return {false, "Missing JSON input"};
+			}
+
+			if (_input->len <= 0) {
+				return {false, "Length of buffer must be greater than zero"};
+			}
+
+			if (!_input->buf) {
+				return {false, "Missing input buffer"};
+			}
+
+			return {true, ""};
+		};
+
+		try {
+			auto* input_bb = boost::any_cast<bytesBuf_t*>(*std::next(std::begin(_rule_arguments), 2));
+			auto& rei = get_rei(_effect_handler);
+			const auto config = load_plugin_config(rei);
+
+			if (const auto [valid, msg] = is_input_valid(input_bb); !valid) {
+				log_re::error(msg);
+				return ERROR(INPUT_ARG_NOT_WELL_FORMED_ERR, msg);
+			}
+
+			const auto input_bb_casted = static_cast<const char*>(input_bb->buf);
+			json input = json::parse(input_bb_casted, input_bb_casted + input_bb->len);
+
+			const auto& operations = input.at("operations");
+
+			for (auto&& op : operations) {
+				const auto& err_code = metadata_guard_handling(op.at("attribute").get_ref<const std::string&>(), rei);
+				if (!err_code.ok()) {
+					return err_code;
+				}
+			}
+
+			return CODE(RULE_ENGINE_CONTINUE);
+		}
+		catch (const json::parse_error& e) {
+			log_re::error({{"log_message", "Failed to parse input into JSON"}, {"error_message", e.what()}});
```

Comment:
Could include the rule_engine_plugin key-value pair here and in the other error cases?